gcs-fetcher/pkg/fetcher/fetcher.go

```diff
@@ -298,10 +320,25 @@ func (gf *Fetcher) fetchObjectOnce(ctx context.Context, j job, dest string, brea
 
 	r, err := gf.GCS.NewReader(ctx, j.bucket, j.object)
 	if err != nil {
+		// Check for AccessDenied failure here and return a useful error message on Stderr and exit immediately.
+		if gerr, ok := err.(*googleapi.Error); ok && gerr.Code == http.StatusForbidden {
+			// Try to parse out the robot name.
+			match := robotRegex.FindStringSubmatch(err.Error())
+			robot := "your Cloud Build robot"
+			if len(match) == 2 {
+				robot = match[1]
+			}
+			gf.logErr("Access to bucket %q denied. You must grant Storage Object Viewer permission to %s.", j.bucket, robot)
+			os.Exit(1)
```

Comment:
It's pretty bad hygiene to have the `fetcher` library `os.Exit`, which shows in the ugliness in the test that has to jump through hoops to catch it. Can this just return an error that bubbles up to `main.go`'s call instead?

Reply:
The key reason for this P-R is to put a short, concise, actionable error message to the user if they get an AccessDenied. The way that Go "wraps" the error message makes this a bit tough. I think it would be worse to remove all the "wrapping context" from the errors.

By "wrapping context", I mean the way that a string is commonly prefixed on the error:
```
  err := something()
  if err != nil {
    return fmt.Errorf("failed to call something: %v", err)
  }
```

Reply:
You don't have to wrap the error to return it, just return err and have code in `main.go` be responsible for `log.Fatal`ing or `os.Exit`ing.

Reply:
What I mean is that there are a lot of layers between this point and main and lots of additional wrapping strings are added along the way. For example:

- https://github.com/GoogleCloudPlatform/cloud-builders/blob/master/gcs-fetcher/pkg/fetcher/fetcher.go#L227

But then to make matters worse, this method doesn't propagate the error and instead is collecting up a bunch of errors because there are up to 200 goroutines working in parallel.

Maybe I just take the performance hit and attempt to read from GCS early (e.g., in main.go) to capture this particular error?

Reply:
main.go doesn't have to be responsible for identifying the specific 403 error, it just has to be responsible for doing the os.Exit when it sees _any_ error returned by `Fetch`.

The work in `processJobs` could probably also be simplified by using an [`ErrGroup`](https://godoc.org/golang.org/x/sync/errgroup) which encapsulates cancelling all the other ongoing goroutines' contexts when one goroutine encounters an error, which could remove a lot of the layers of wrapping and propagating. It seems like that's what you're intending to do by `os.Exit`ing when you encounter the Access Denied, to stop all the other fetcher goroutines.

Reply:
No. If you'd prefer, we can merge and release this PR even with the `os.Exit` in the library, and I can do the work to refactor to use an `ErrGroup`. I'll make sure you get a chance to review and test that PR before we release it.

Reply:
ty

Reply:
In the refactor, the key thing is to get a simple error message written to $BUILDER_OUTPUT/output.

Good:
`Access to bucket XXX denied. You must grant Storage Object Viewer permission to YYY.`

Bad:
`failed to fetch: failed to fetchOnce: failed to fetchOnceWithTimeout: Access to bucket XXX denied. You must grant Storage Object Viewer permission to YYY.`